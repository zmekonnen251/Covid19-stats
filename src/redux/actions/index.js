import { LOAD_DATA, SET_CONTINENT, SET_COUNTRY } from '../actionTypes';

import fetchAllApi from '../../api';
import loadDataByContinent from '../../modules/continent';

export const loadData = () => async (dispatch) => {
  const response = await fetchAllApi();
  const { Global: globalData, Countries: countries } = response.data;
  const date = response.data.Date.split('T')[0];

  const continentsData = loadDataByContinent(countries);
  dispatch({ type: LOAD_DATA, payload: { globalData, continentsData, date } });
};

export const setContinent = (continent = '') => ({
  type: SET_CONTINENT,
  payload: { continent },
});

export const setCountry = (country = '') => ({
  type: SET_COUNTRY,
  payload: { country },
});
