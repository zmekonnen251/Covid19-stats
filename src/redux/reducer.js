import axios from 'axios';
import getContinent from '../modules/continent';
import { LOAD_DATA, SET_CONTINENT, SET_COUNTRY } from './actionTypes';

const covidInitial = { date: '', continents: {} };
export const covid19Reducer = (state = covidInitial, action = {}) => {
  switch (action.type) {
    case LOAD_DATA:
      return { ...state, ...action.payload };
    default:
      return state;
  }
};

export const continentReducer = (state = {}, action = {}) => {
  switch (action.type) {
    case SET_CONTINENT:
      return action.payload;
    default:
      return state;
  }
};

export const countryReducer = (state = {}, action = {}) => {
  switch (action.type) {
    case SET_COUNTRY:
      return action.payload;
    default:
      return state;
  }
};

export const loadData = (selectedDate = '') => {
  const date = new Date();
  const urlDate = selectedDate || date.toISOString().slice(0, 10);
  const url = `https://api.covid19tracking.narrativa.com//api/${urlDate}`;
  const continent = {
    date: '',
    continents: {
      Africa: { name: 'Africa', countries: {} },
      Europe: { name: 'Europe', countries: {} },
      North_America: { name: 'North_America', countries: {} },
      Oceania: { name: 'Oceania', countries: {} },
      Asia: { name: 'Asia', countries: {} },
      South_America: { name: 'South_America', countries: {} },
    },
  };

  return async (dispatch) => {
    const response = await axios.get(url);
    const dataWithSpecificDate = response.data.dates;

    const newDate = Object.keys(dataWithSpecificDate)[0];

    const finalData = dataWithSpecificDate[newDate];
    continent.date = newDate;
    Object.keys(finalData.countries).forEach((c) => {
      const country = finalData.countries[c];
      switch (getContinent(country.name.split('_').join(' '))) {
        case 'Europe':
          continent.continents.Europe.countries[country.name] = {
            name: country.name,
            today_confirmed: country.today_confirmed,
            today_deaths: country.today_deaths,
            today_new_confirmed: country.today_new_confirmed,
            today_new_deaths: country.today_new_deaths,
            today_new_open_cases: country.today_new_open_cases,
            today_new_recovered: country.today_new_recovered,
            today_open_cases: country.today_open_cases,
            today_recovered: country.today_recovered,
          };
          break;
        case 'Africa':
          continent.continents.Africa.countries[country.name] = {
            name: country.name,
            today_confirmed: country.today_confirmed,
            today_deaths: country.today_deaths,
            today_new_confirmed: country.today_new_confirmed,
            today_new_deaths: country.today_new_deaths,
            today_new_open_cases: country.today_new_open_cases,
            today_new_recovered: country.today_new_recovered,
            today_open_cases: country.today_open_cases,
            today_recovered: country.today_recovered,
          };
          break;
        case 'Asia':
          continent.continents.Asia.countries[country.name] = {
            name: country.name,
            today_confirmed: country.today_confirmed,
            today_deaths: country.today_deaths,
            today_new_confirmed: country.today_new_confirmed,
            today_new_deaths: country.today_new_deaths,
            today_new_open_cases: country.today_new_open_cases,
            today_new_recovered: country.today_new_recovered,
            today_open_cases: country.today_open_cases,
            today_recovered: country.today_recovered,
          };
          break;
        case 'North_America':
          continent.continents.North_America.countries[country.name] = {
            name: country.name,
            today_confirmed: country.today_confirmed,
            today_deaths: country.today_deaths,
            today_new_confirmed: country.today_new_confirmed,
            today_new_deaths: country.today_new_deaths,
            today_new_open_cases: country.today_new_open_cases,
            today_new_recovered: country.today_new_recovered,
            today_open_cases: country.today_open_cases,
            today_recovered: country.today_recovered,
          };
          break;
        case 'South_America':
          continent.continents.South_America.countries[country.name] = {
            name: country.name,
            today_confirmed: country.today_confirmed,
            today_deaths: country.today_deaths,
            today_new_confirmed: country.today_new_confirmed,
            today_new_deaths: country.today_new_deaths,
            today_new_open_cases: country.today_new_open_cases,
            today_new_recovered: country.today_new_recovered,
            today_open_cases: country.today_open_cases,
            today_recovered: country.today_recovered,
          };
          break;
        case 'Oceania':
          continent.continents.Oceania.countries[country.name] = {
            name: country.name,
            today_confirmed: country.today_confirmed,
            today_deaths: country.today_deaths,
            today_new_confirmed: country.today_new_confirmed,
            today_new_deaths: country.today_new_deaths,
            today_new_open_cases: country.today_new_open_cases,
            today_new_recovered: country.today_new_recovered,
            today_open_cases: country.today_open_cases,
            today_recovered: country.today_recovered,
          };
          break;
        default:
          break;
      }
    });

    dispatch({ type: LOAD_DATA, payload: continent });
  };
};

export const setContinent = (continent = '') => ({
  type: SET_CONTINENT,
  payload: { continent },
});

export const setCountry = (country = '') => ({
  type: SET_COUNTRY,
  payload: { country },
});
