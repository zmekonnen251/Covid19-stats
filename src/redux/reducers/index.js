import { LOAD_DATA, SET_CONTINENT, SET_COUNTRY } from '../actionTypes';

const covidInitial = {
  date: '', continentsData: {}, globalData: {}, selectedContinent: null, selectedCountry: null,
};

export default (state = covidInitial, action = {}) => {
  switch (action.type) {
    case LOAD_DATA:
      return { ...state, ...action.payload };
    case SET_CONTINENT:
      return { ...state, selectedContinent: action.payload };
    case SET_COUNTRY:
      return { ...state, selectedCountry: action.payload };
    default:
      return state;
  }
};
