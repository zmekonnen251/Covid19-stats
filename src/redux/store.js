import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunk from 'redux-thunk';
import covidReducer from './reducers';

const rootReducer = combineReducers({
  covidData: covidReducer,
});

const store = createStore(rootReducer, applyMiddleware(thunk));

export default store;
