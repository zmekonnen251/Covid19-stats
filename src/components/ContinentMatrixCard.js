import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { BsFillArrowDownRightSquareFill } from 'react-icons/bs';
import { setContinent } from '../redux/reducer';

const ContinentMatrixCard = (props) => {
  const dispatch = useDispatch();
  const [continent, setContinentLocal] = useState('');
  const {
    name, totalConfirmed, totalDeath, onClickSetContinent,
  } = props;

  const continentMatrixClickHandler = () => {
    setContinentLocal(onClickSetContinent);
  };

  useEffect(() => {
    dispatch(setContinent(continent));
  }, [continent]);

  return (
    <>
      <div
        className="relative border p-8 cursor-pointer flex flex-col items-center gap-1"
        role="presentation"
        onClick={continentMatrixClickHandler}
      >
        <h1 className="text-4xl">{name}</h1>
        <h3 className="text-lg">
          Total Confirmed Cases :
          {totalConfirmed}
        </h3>
        <h3 className="text-lg">
          Total Deaths :
          {totalDeath}
        </h3>
        <div className="absolute bottom-1 right-1 flex justify-end">
          <BsFillArrowDownRightSquareFill />
        </div>
      </div>
    </>
  );
};

ContinentMatrixCard.propTypes = {
  name: PropTypes.string.isRequired,
  totalConfirmed: PropTypes.number.isRequired,
  totalDeath: PropTypes.number.isRequired,
  onClickSetContinent: PropTypes.string.isRequired,
};

export default ContinentMatrixCard;
