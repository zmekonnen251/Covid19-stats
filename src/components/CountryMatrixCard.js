import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';
import { BsFillArrowDownRightSquareFill } from 'react-icons/bs';
import { setCountry } from '../redux/reducer';

const CountryMatrixCard = (props) => {
  const dispatch = useDispatch();
  const {
    name,
    totalConfirmed,
    totalDeath,
    onClickSetCountry,
    selectedContinent,
  } = props;
  const [country, setLocalCountry] = useState('');

  const countryMatrixClickHandler = () => {
    setLocalCountry(onClickSetCountry);
  };

  useEffect(() => {
    dispatch(setCountry(country));
  }, [country]);

  return (
    <>
      <NavLink
        to={`/continent/${selectedContinent}/details/${name
          .split(' ')
          .join('_')}`}
        onClick={countryMatrixClickHandler}
        className="relative opacity-70 rounded-lg hover:border bg-pink-800 hover:bg-pink-700 hover:w-[99%] h-[200px] w-full  p-16 cursor-pointer flex flex-col justify-between items-end gap-6"
        role="presentation"
      >
        <div
          className="absolute box-border left-[-170px] inset-1 bottom-8  z-0 opacity-20"
          style={{
            backgroundImage: `url(https://mapsvg.com/static/maps/geo-calibrated/${name
              .split('_')
              .join('-')
              .toLowerCase()}.svg)`,
            backgroundSize: '15%',
            backgroundRepeat: 'no-repeat',
            backgroundPosition: 'center',
          }}
        />
        <div className="absolute right-1 bottom-4">
          <h1 className="text-2xl">{name}</h1>
          <h3 className="text-lg">
            Total Confirmed Cases :
            {` ${totalConfirmed}`}
          </h3>
          <h3 className="text-base">
            Total Deaths :
            {` ${totalDeath}`}
          </h3>
          <div className="absolute bottom-1 right-1 flex justify-end">
            <BsFillArrowDownRightSquareFill />
          </div>
        </div>
      </NavLink>
    </>
  );
};

CountryMatrixCard.propTypes = {
  name: PropTypes.string.isRequired,
  totalConfirmed: PropTypes.number.isRequired,
  totalDeath: PropTypes.number.isRequired,
  selectedContinent: PropTypes.string.isRequired,
  onClickSetCountry: PropTypes.string.isRequired,
};

export default CountryMatrixCard;
