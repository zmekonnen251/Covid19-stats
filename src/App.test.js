import renderer from 'react-test-renderer';
import { cleanup } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import {
  covid19Reducer,
  continentReducer,
  countryReducer,
} from './redux/reducer';

import Main from './components/Main';
import CountryMatrixCard from './components/CountryMatrixCard';
import ContinentMatrixCard from './components/ContinentMatrixCard';

const rootReducer = combineReducers({
  covidData: covid19Reducer,
  continent: continentReducer,
  country: countryReducer,
});
const store = createStore(rootReducer, applyMiddleware(thunk));

afterEach(cleanup);

describe('Home Page', () => {
  test('Home page renders correctly', () => {
    const homePage = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Main />
          </Router>
        </Provider>
      )
      .toJSON();
    expect(homePage).toMatchSnapshot();
  });
});

describe('Countries List ', () => {
  test('Countries Card renders correctly', () => {
    const countryCard = renderer
      .create(
        <Provider store={store}>
          <Router>
            <CountryMatrixCard
              name="Ethiopia"
              totalConfirmed={20000}
              totalDeath={1000}
              onClickSetCountry="Ethiopia"
            />
          </Router>
        </Provider>
      )
      .toJSON();
    expect(countryCard).toMatchSnapshot();
  });
});

describe('Countries List ', () => {
  test('Continent Card renders correctly', () => {
    const continentCard = renderer
      .create(
        <Provider store={store}>
          <Router>
            <ContinentMatrixCard
              key={uuidv4()}
              name="Europe"
              totalConfirmed={500000}
              totalDeath={50000}
              onClickSetContinent={'Europe'}
            />
            ;
          </Router>
        </Provider>
      )
      .toJSON();
    expect(continentCard).toMatchSnapshot();
  });
});
