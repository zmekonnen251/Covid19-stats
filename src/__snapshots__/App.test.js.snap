// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries List  Continent Card renders correctly 1`] = `
Array [
  <div
    className="relative border p-8 cursor-pointer flex flex-col items-center gap-1"
    onClick={[Function]}
    role="presentation"
  >
    <h1
      className="text-4xl"
    >
      Europe
    </h1>
    <h3
      className="text-lg"
    >
      Total Confirmed Cases :
      500000
    </h3>
    <h3
      className="text-lg"
    >
      Total Deaths :
      50000
    </h3>
    <div
      className="absolute bottom-1 right-1 flex justify-end"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14 16a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12zM5.904 5.197 10 9.293V6.525a.5.5 0 0 1 1 0V10.5a.5.5 0 0 1-.5.5H6.525a.5.5 0 0 1 0-1h2.768L5.197 5.904a.5.5 0 0 1 .707-.707z"
        />
      </svg>
    </div>
  </div>,
  ";",
]
`;

exports[`Countries List  Countries Card renders correctly 1`] = `
<div
  className="relative border p-8 cursor-pointer flex flex-col items-center gap-1"
  onClick={[Function]}
  role="presentation"
>
  <h1
    className="text-3xl"
  >
    Ethiopia
  </h1>
  <h3
    className="text-lg"
  >
    Total Confirmed Cases :
    20000
  </h3>
  <h3
    className="text-lg"
  >
    Total Deaths :
    1000
  </h3>
  <div
    className="absolute bottom-1 right-1 flex justify-end"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M14 16a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12zM5.904 5.197 10 9.293V6.525a.5.5 0 0 1 1 0V10.5a.5.5 0 0 1-.5.5H6.525a.5.5 0 0 1 0-1h2.768L5.197 5.904a.5.5 0 0 1 .707-.707z"
      />
    </svg>
  </div>
</div>
`;

exports[`Home Page Home page renders correctly 1`] = `
Array [
  <h1
    className="text-3xl font-bold underline text-center"
  >
    Covid Stats
  </h1>,
  <form
    className="text-center  m-2"
  >
    <input
      className="bg-inherit text-inherit border p-1 mt-2"
      onChange={[Function]}
      type="date"
    />
  </form>,
  <div
    className="grid grid-cols-1 sm:grid-cols-2 gap-1 m-7 border-box"
  >
    <div
      className="sm:col-span-2"
      role="presentation"
    >
      <h1
        className="text-4xl mb-3 text-center"
      >
        World Total
      </h1>
      <h1
        className="text-2xl mb-3 text-center"
      />
      <div
        className="border p-8 cursor-pointer flex flex-col items-center gap-1"
      >
        <h3
          className="text-lg"
        >
          Total Confirmed Cases :
          0
        </h3>
        <h3
          className="text-lg"
        >
          Total Deaths :
          0
        </h3>
        <h3
          className="text-lg"
        >
          Total Recovered :
          0
        </h3>
        <h3
          className="text-lg"
        >
          Total Open Cases :
          0
        </h3>
      </div>
    </div>
  </div>,
]
`;
